api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java

```diff
@@ -203,4 +275,24 @@ private ProjectionUtil() {}
         return null;
     }
   }
+
+  static <S, T> UnboundPredicate<T> truncateArrayStrict(
+      String name, BoundPredicate<S> pred, Transform<S, T> transform) {
+    S boundary = pred.literal().value();
+    switch (pred.op()) {
+      case LT:
+      case LT_EQ:
+        return predicate(Expression.Operation.LT, name, transform.apply(boundary));
+      case GT:
+      case GT_EQ:
+        return predicate(Expression.Operation.GT, name, transform.apply(boundary));
+      case NOT_EQ:
+        return predicate(Expression.Operation.NOT_EQ, name, transform.apply(boundary));
+      case EQ:
+        // there is no predicate that guarantees equality because adjacent values transform to the same partition
+        return null;
```

Comment:
If the length of the array is less than the truncate length, would the original predicate work?

Reply:
Ah, let me check. Will add a test for that.

Reply:
Ah so if the array length is less than the truncate length, i.e. `"value" == transform("value")`, it's just going to be a single value partition. Then yes, it can accurately return the original predicate for all the cases including `EQ`.

Will add that case and corresponding tests.

Reply:
Have updated the cases for both strict and inclusive projection. 

Since array/string length is less than the truncate length, it's always going to be a partition containing that single value. So, for all cases it can always reflect the original predicate. Does that make sense?

Reply:
(Corresponding tests: https://github.com/apache/incubator-iceberg/pull/332/files#diff-f4be406e8a0e2a766149e6c9fe56142fR273
https://github.com/apache/incubator-iceberg/pull/332/files#diff-f4be406e8a0e2a766149e6c9fe56142fR309)

Reply:
Reverted the ^ change. Will return `null` for `EQ` and keep rest of the logic same.